src/python/pants/backend/native/config/environment.py

```diff
@@ -166,15 +163,95 @@ def get_invocation_environment_dict(self, platform):
 class GCCCCompiler(datatype([('c_compiler', CCompiler)])): pass
 
 
-class LLVMCCompiler(datatype([('c_compiler', CCompiler)])): pass
+class GCCCLinker(datatype([('c_linker', Linker)])): pass
 
 
 class GCCCppCompiler(datatype([('cpp_compiler', CppCompiler)])): pass
 
 
+class GCCCppLinker(datatype([('cpp_linker', Linker)])): pass
+
+
+class LLVMCCompiler(datatype([('c_compiler', CCompiler)])): pass
+
+
+class LLVMCLinker(datatype([('c_linker', Linker)])): pass
+
+
 class LLVMCppCompiler(datatype([('cpp_compiler', CppCompiler)])): pass
 
 
+class LLVMCppLinker(datatype([('cpp_linker', Linker)])): pass
+
+
+class CToolchain(datatype([('c_compiler', CCompiler), ('c_linker', Linker)])): pass
+
+
+class CToolchainProvider(object):
+
+  @abstractproperty
+  def as_c_toolchain(self):
+    """???"""
+
+
+class LLVMCToolchain(datatype([
+    ('llvm_c_compiler', LLVMCCompiler),
+    ('llvm_c_linker', LLVMCLinker),
+]), CToolchainProvider):
```

Comment:
I'm not understanding why `class LLVMCCompiler(datatype([('c_compiler', CCompiler)])): pass` vs just `class LLVMCCompiler(CCompiler): pass`, etc. It seems to me this would eliminate the need for both `CToolchainProvider` and `CppToolchainProvider`.

Reply:
You're completely right. 3441e74 does exactly this. That removed ~70 lines from this diff.